mozilla/bedrock
# Add SSO auth to Wagtail and Django admins

## One-line summary

We're not exposing the Wagtail or Django admins in production yet (and there's nothing to see yet anyway), but before we do, we need to ensure auth is only possible via Mozilla SSO. This changeset delivers that.

- [ ] I used an AI to write some of this code.

## Significant changes and points to review

The changes here basically involve plugging in our OIDC library `mozilla-django-oidc` and updating the login templates to show a "Sign in with Mozilla SSO" button instead of the regular username + password fields. That looks like this:

<img width="524" alt="Screenshot 2024-06-05 at 19 54 21" src="https://github.com/mozilla/bedrock/assets/101457/407beaef-9d2e-44e6-87cb-fa004b31cf99">

<img width="549" alt="Screenshot 2024-06-05 at 19 54 09" src="https://github.com/mozilla/bedrock/assets/101457/07e5f1e0-026e-4c8d-a11b-df852cf00115">

(The footnote about the redirect there is not ideal, but given the low usage of the Django admin compared to the Wagtail admin, I don't think it's worth sinking in a load of time on addressing that)

Also, one rough patch with using OIDC is that session leases can expire while someone has a page open - particularly if they leave a draft in progress over a day, say. (This will trigger a CSRF error because the auto-renewed lease will have a different token associated with it). I've pushed up the default session lease time (`OIDC_RENEW_ID_TOKEN_EXPIRY_SECONDS`) to 18 hours, vs the library's default of an hour.

Also, lifted from Birdbox, we have a custom CSRF error page (rather than Django's yellow, debug-like one)

Logged out users (who are very unlikely to see this anyway) get a simple version:
<img width="906" alt="csrf1" src="https://github.com/mozilla/bedrock/assets/101457/97fb044e-9c87-43a8-8184-9387fbb40632">

Logged in, staff/CMS users get a bit more context:
<img width="857" alt="csrf2" src="https://github.com/mozilla/bedrock/assets/101457/a9f4122f-a19e-4c43-90ae-fba388a58834">


## Issue / Bugzilla link

Resolves #14641

## Testing

* Check out this branch
* Extend your `.env` file with the development OIDC credentials from our password vault - search for "Bedrock CMS SSO" and pick the `bedrock-localdev` ones at the bottom. If you need a `.env` file, copy `.env-dist` then add the credentials
* Ensure your `.env` entry of `USE_SSO_AUTH` is `True` (or absent entirely, because that's the default value)
* Create a local admin user with `./manage.py createsuperuser` ensuring you set the username AND email to your Mozilla SSO/LDAP identity's address. You can set the password to garbage - it will get never get used and gets wiped once SSO is used (IIRC).
* Note that all of the Websites team have been _invited_ to the appropriate people.m.o access group (`bedrock_cms_local_dev-access`) - you need to accept the invitation before you can log in
* Start bedrock running (`npm start` or `make build run` for Docker)
* Go to `/cms-admin/` and you should see a button to login with SSO. Click it and you should go through the OAuth flow and end up in the Wagtail admin (because you're an admin).
* Sign out from bottom right
* Go to /django-admin/ and confirm you can sign in there with SSO, too, then sign out
* Go to /django-rq/ (the task queue UI) and confirm you can sign in there with SSO, too
